src/plugins/interactive_setup/server/routes/enroll.ts

```diff
@@ -6,26 +6,110 @@
-        body: schema.object({ token: schema.string() }),
+        body: schema.object({
+          hosts: schema.arrayOf(schema.uri({ scheme: 'https' }), {
+            minSize: 1,
+          }),
+          apiKey: schema.string(),
+          rootCAFingerprint: schema.maybe(schema.string()),
+        }),
       },
       options: { authRequired: false },
     },
     async (context, request, response) => {
-      return response.forbidden({
-        body: { message: `API is not implemented yet.` },
-      });
+      if (!preboot.isSetupOnHold()) {
+        logger.error(`Invalid request to [path=${request.url.pathname}] outside of preboot phase`);
+        return response.badRequest();
+      }
+
+      const connectionStatus = await elasticsearch.connectionStatus$.pipe(first()).toPromise();
+      if (connectionStatus === ElasticsearchConnectionStatus.Configured) {
+        logger.error(
+          `Invalid request to [path=${request.url.pathname}], Elasticsearch connection is already configured.`
+        );
+        return response.badRequest({
+          body: {
+            message: 'Elasticsearch connection is already configured.',
+            attributes: { type: 'elasticsearch_connection_configured' },
+          },
+        });
+      }
+
+      // The most probable misconfiguration case is when Kibana process isn't allowed to write to the
+      // Kibana configuration file. We'll still have to handle possible filesystem access errors
+      // when we actually write to the disk, but this preliminary check helps us to avoid unnecessary
+      // enrollment call and communicate that to the user early.
+      const isConfigWritable = await kibanaConfig.isConfigWritable();
+      if (!isConfigWritable) {
+        logger.error('Kibana process does not have enough permissions to write to config file');
+        return response.customError({
+          statusCode: 500,
+          body: {
+            message: 'Kibana process does not have enough permissions to write to config file.',
+            attributes: { type: 'kibana_config_not_writable' },
+          },
+        });
+      }
+
+      let enrollResult: EnrollResult;
+      try {
+        enrollResult = await elasticsearch.enroll({
+          apiKey: request.body.apiKey,
+          hosts: request.body.hosts,
+          caFingerprint: request.body.rootCAFingerprint,
+        });
+      } catch {
+        // For security reasons, we shouldn't leak to the user whether Elasticsearch node couldn't process enrollment
+        // request or we just couldn't connect to any of the provided hosts.
+        return response.customError({
+          statusCode: 500,
+          body: { message: 'Failed to enroll.', attributes: { type: 'enroll_failure' } },
+        });
+      }
+
+      try {
+        await kibanaConfig.writeConfig({
+          host: enrollResult.host,
+          ca: enrollResult.ca,
+          username: enrollResult.username,
+          password: enrollResult.password,
+        });
```

Comment:
nit: Can we just pass in `enrollResult` as a whole instead of picking individual properties?

Reply:
Yeah, just wasn't initially sure if all fields matches.